listenbrainz_spark/stats/user.py

```diff
@@ -0,0 +1,190 @@
+            'track_name': row.track_name,
+            'recording_msid': row.recording_msid,
+            'recording_mbid': row.recording_mbid,
+            'artist_name': row.artist_name,
+            'artist_msid': row.artist_msid,
+            'artist_mbids': row.artist_mbids,
+            'release_name': row.release_name,
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'listen_count': row.cnt,
+        })
+    query_t0 = time.time()
+    print("Query to calculate artist stats processed in %.2f s" % (query_t0 - t0))
+
+def get_releases(table):
+    """
+    Args:
+        table: name of the temporary table
+    """
+    t0 = time.time()
+    query = run_query("""
+            SELECT user_name
+                 , release_name
+                 , release_msid
+                 , release_mbid
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(release_msid) as cnt
+              FROM %s
+          GROUP BY user_name, release_name, release_msid, release_mbid, artist_name, artist_msid, artist_mbids
+          ORDER BY cnt DESC
+        """ % (table))
+    releases = defaultdict(list)
+    t = time.time()
+    rows = query.collect()
+    releases = defaultdict(list)
+    print("time taken by collect call: %.2f" % (time.time() - t))
+    for row in rows:
+        releases[row.user_name].append({
+            'release_name': row.release_name,
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'artist_name': row.artist_name,
+            'artist_msid': row.artist_msid,
+            'artist_mbids': row.artist_mbids,
+            'listen_count': row.cnt,
+        })
+    query_t0 = time.time()
+    print("Query to calculate artist stats processed in %.2f s" % (query_t0 - t0))
+
+
+def get_users(table):
+    """ DataFrame is registered as a temporary table
+    Args:
+        table : name of the temporary table
+
+    Returns:
+        users: list containing user names of all the users
+    """
+    t0 = time.time()
+    query = run_query("""
+            SELECT DISTINCT user_name
+              FROM %s
+        """ % table)
+    users = [row.user_name for row in query.collect()]
+    query_t0 = time.time()
+    query.show()
+    print("Query to get list of users processed in %.2f s" % (query_t0 - t0))
+    return users
+
+def main(app_name):
+    """
+    Args:
+        app_name: Application name to uniquely identify the submitted
+                  application amongst other applications
+    """
+    t0 = time.time()
+    listenbrainz_spark.init_spark_session(app_name)
+    t = datetime.utcnow().replace(day=1)
+    date = t + relativedelta(months=-1)
+    try:
+        df = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'.format(config.HDFS_CLUSTER_URI, date.year, date.month))
+        print("Loading dataframe...")
+    except:
```

Comment:
A blanket exception catch is not a good idea in 99% of cases -- it often hides other exceptions, as it does in this case. If the script cannot read the parqet file, it will happily tell us that there are no new listens. Whoops.